src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -284,6 +334,48 @@ end
 @[simp] lemma length_eq_zero_iff {u : V} {p : G.walk u u} : p.length = 0 ↔ p = nil :=
 by cases p; simp
 
+section concat_rec
+
+variables
+  {motive : Π (u v : V), G.walk u v → Sort*}
+  (Hnil : Π {u : V}, motive u u nil)
+  (Hconcat : Π {u v w : V} (p : G.walk u v) (h : G.adj v w), motive u v p → motive u w (p.concat h))
+
+/-- Auxiliary definition for `simple_graph.walk.concat_rec` -/
+def concat_rec_aux : Π {u v : V} (p : G.walk u v), motive v u p.reverse
+| _ _ nil := Hnil
+| _ _ (cons h p) := eq.rec (Hconcat p.reverse (G.symm h) (concat_rec_aux p)) (reverse_cons h p).symm
+
+lemma concat_rec_aux_concat {u v w : V} (p : G.walk u v) (h : G.adj v w) :
+  concat_rec_aux @Hnil @Hconcat (p.concat h).reverse
+  == Hconcat p.reverse.reverse h (concat_rec_aux @Hnil @Hconcat p.reverse) :=
+begin
+  transitivity concat_rec_aux @Hnil @Hconcat (cons h.symm p.reverse),
+  { congr, simp },
+  simp [concat_rec_aux, rec_heq_iff_heq],
+end
+
+/-- Recursor on walks by inducting on `simple_graph.walk.concat`. -/
+@[elab_as_eliminator]
+def concat_rec {u v : V} (p : G.walk u v) : motive u v p :=
```

Comment:
Can you link to the corresponding `cons` recursor (and back if it's not autogenerated)?